internal/mempool/v1/mempool.go

```diff
@@ -716,33 +667,45 @@ func (txmp *TxMempool) recheckTransactions() {
 		"num_txs", txmp.Size(),
 		"height", txmp.height,
 	)
-	// N.B.: We have to issue the calls outside the lock. In a local client,
-	// even an "async" call invokes its callback immediately which will make the
-	// callback deadlock trying to acquire the same lock. This isn't a problem
-	// with out-of-process calls, but this has to work for both.
-	txmp.mtx.Unlock()
-	defer txmp.mtx.Lock()
 
-	ctx := context.TODO()
-	atomic.StoreInt64(&txmp.txRecheck, int64(txmp.txs.Len()))
+	// Collect transactions currently in the mempool requiring recheck.
+	wtxs := make([]*WrappedTx, 0, txmp.txs.Len())
 	for e := txmp.txs.Front(); e != nil; e = e.Next() {
-		wtx := e.Value.(*WrappedTx)
-
-		// The response for this CheckTx is handled by the default recheckTxCallback.
-		_, err := txmp.proxyAppConn.CheckTxAsync(ctx, abci.RequestCheckTx{
-			Tx:   wtx.tx,
-			Type: abci.CheckTxType_Recheck,
-		})
-		if err != nil {
-			txmp.logger.Error("failed to execute CheckTx during recheck",
-				"err", err, "hash", fmt.Sprintf("%x", wtx.tx.Hash()))
-			atomic.AddInt64(&txmp.txRecheck, -1)
+		wtxs = append(wtxs, e.Value.(*WrappedTx))
+	}
+
+	// Issue CheckTx calls for each remaining transaction, and when all the
+	// rechecks are complete signal watchers that transactions may be available.
+	go func() {
+		ctx := context.TODO()
+		g, start := taskgroup.New(nil).Limit(2 * runtime.NumCPU())
+
+		for _, wtx := range wtxs {
+			wtx := wtx
+			start(func() error {
+				// The response for this CheckTx is handled by the default recheckTxCallback.
+				rsp, err := txmp.proxyAppConn.CheckTxSync(ctx, abci.RequestCheckTx{
+					Tx:   wtx.tx,
+					Type: abci.CheckTxType_Recheck,
+				})
+				if err != nil {
+					txmp.logger.Error("failed to execute CheckTx during recheck",
+						"err", err, "hash", fmt.Sprintf("%x", wtx.tx.Hash()))
+				}
+				txmp.recheckTxCallback(wtx.tx, rsp)
```

Comment:
From conversation: Should we skip this on error?

Reply:
Done. I also renamed the callbacks (which are no longer callbacks) as we discussed.